{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "a830bf8d",
   "metadata": {},
   "source": [
    "# Conversion to other organisms\n",
    "\n",
    "Most of the prior knowledge stored inside `Omnipath` is derived from human data, therefore they use gene names. Despite this, using homology we can convert gene names to other organisms.\n",
    "\n",
    "To showcase how to do it inside `decoupler`, we will load the `MSigDB` database and convert it into gene symbols for mouse and fly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b70696b1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>MSC</td>\n",
       "      <td>oncogenic_signatures</td>\n",
       "      <td>PKCA_DN.V1_DN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>MSC</td>\n",
       "      <td>mirna_targets</td>\n",
       "      <td>MIR12123</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>MSC</td>\n",
       "      <td>chemical_and_genetic_perturbations</td>\n",
       "      <td>NIKOLSKY_BREAST_CANCER_8Q12_Q22_AMPLICON</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>MSC</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>GSE32986_UNSTIM_VS_GMCSF_AND_CURDLAN_LOWDOSE_S...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>MSC</td>\n",
       "      <td>chemical_and_genetic_perturbations</td>\n",
       "      <td>BENPORATH_PRC2_TARGETS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407729</th>\n",
       "      <td>OR2W5P</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>GSE22601_DOUBLE_NEGATIVE_VS_CD8_SINGLE_POSITIV...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407730</th>\n",
       "      <td>OR2W5P</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>KANNAN_BLOOD_2012_2013_TIV_AGE_65PLS_REVACCINA...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407731</th>\n",
       "      <td>OR52L2P</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>GSE22342_CD11C_HIGH_VS_LOW_DECIDUAL_MACROPHAGE...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407732</th>\n",
       "      <td>CSNK2A3</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>OCONNOR_PBMC_MENVEO_ACWYVAX_AGE_30_70YO_7DY_AF...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2407733</th>\n",
       "      <td>AQP12B</td>\n",
       "      <td>immunologic_signatures</td>\n",
       "      <td>MATSUMIYA_PBMC_MODIFIED_VACCINIA_ANKARA_VACCIN...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2407734 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        genesymbol                          collection  \\\n",
       "0              MSC                oncogenic_signatures   \n",
       "1              MSC                       mirna_targets   \n",
       "2              MSC  chemical_and_genetic_perturbations   \n",
       "3              MSC              immunologic_signatures   \n",
       "4              MSC  chemical_and_genetic_perturbations   \n",
       "...            ...                                 ...   \n",
       "2407729     OR2W5P              immunologic_signatures   \n",
       "2407730     OR2W5P              immunologic_signatures   \n",
       "2407731    OR52L2P              immunologic_signatures   \n",
       "2407732    CSNK2A3              immunologic_signatures   \n",
       "2407733     AQP12B              immunologic_signatures   \n",
       "\n",
       "                                                   geneset  \n",
       "0                                            PKCA_DN.V1_DN  \n",
       "1                                                 MIR12123  \n",
       "2                 NIKOLSKY_BREAST_CANCER_8Q12_Q22_AMPLICON  \n",
       "3        GSE32986_UNSTIM_VS_GMCSF_AND_CURDLAN_LOWDOSE_S...  \n",
       "4                                   BENPORATH_PRC2_TARGETS  \n",
       "...                                                    ...  \n",
       "2407729  GSE22601_DOUBLE_NEGATIVE_VS_CD8_SINGLE_POSITIV...  \n",
       "2407730  KANNAN_BLOOD_2012_2013_TIV_AGE_65PLS_REVACCINA...  \n",
       "2407731  GSE22342_CD11C_HIGH_VS_LOW_DECIDUAL_MACROPHAGE...  \n",
       "2407732  OCONNOR_PBMC_MENVEO_ACWYVAX_AGE_30_70YO_7DY_AF...  \n",
       "2407733  MATSUMIYA_PBMC_MODIFIED_VACCINIA_ANKARA_VACCIN...  \n",
       "\n",
       "[2407734 rows x 3 columns]"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import decoupler as dc\n",
    "\n",
    "msigdb = dc.get_resource('MSigDB')\n",
    "msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "50ab6a57",
   "metadata": {},
   "source": [
    "For this example we will filter by the `hallmark` gene sets collection:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ad44cfa4",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>MSC</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>149</th>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>223</th>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_INFLAMMATORY_RESPONSE</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>270</th>\n",
       "      <td>ICOSLG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_ALLOGRAFT_REJECTION</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>398</th>\n",
       "      <td>FOSL2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_HYPOXIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878342</th>\n",
       "      <td>FOXO1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878418</th>\n",
       "      <td>GCG</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878512</th>\n",
       "      <td>PDX1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878605</th>\n",
       "      <td>INS</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>878785</th>\n",
       "      <td>SRP9</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7318 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genesymbol collection                           geneset\n",
       "11            MSC   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "149        ICOSLG   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "223        ICOSLG   hallmark    HALLMARK_INFLAMMATORY_RESPONSE\n",
       "270        ICOSLG   hallmark      HALLMARK_ALLOGRAFT_REJECTION\n",
       "398         FOSL2   hallmark                  HALLMARK_HYPOXIA\n",
       "...           ...        ...                               ...\n",
       "878342      FOXO1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878418        GCG   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878512       PDX1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878605        INS   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "878785       SRP9   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "\n",
       "[7318 rows x 3 columns]"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filter by hallmark\n",
    "msigdb = msigdb[msigdb['collection']=='hallmark']\n",
    "\n",
    "# Remove duplicated entries\n",
    "msigdb = msigdb[~msigdb.duplicated(['geneset', 'genesymbol'])]\n",
    "msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0f3ac365",
   "metadata": {},
   "source": [
    "Then, we can easily transform the obtained resource into mouse genes. For that we will use the NCBI taxonomy identifiers, we go from human (`9606`) to mouse (`10090`). If you want to convert genes into another model organism, you can check its corresponding id [here](https://www.ncbi.nlm.nih.gov/taxonomy).\n",
    "\n",
    "<div class=\"alert alert-info\">\n",
    "\n",
    "**Note**\n",
    "    \n",
    "The first time using this function might take a while (~ 15 minutes). Since the data is stored in cache, the next times is going to run faster. If you need to reset the cache, run `rm -r .pypath/cache/`.\n",
    "\n",
    "</div>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "819128fd",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[2022-11-25 13:46:39] [curl] Module `pysftp` not available. Only downloading of a small number of resources relies on this module. Please install by PIP if it is necessary for you.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Msc</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Fosl2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_HYPOXIA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Fosl2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Relb</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Plau</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7684</th>\n",
       "      <td>Gcg</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7685</th>\n",
       "      <td>Pdx1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7686</th>\n",
       "      <td>Ins1</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7687</th>\n",
       "      <td>Ins2</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7688</th>\n",
       "      <td>Srp9</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7551 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "     genesymbol collection                           geneset\n",
       "0           Msc   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "1         Fosl2   hallmark                  HALLMARK_HYPOXIA\n",
       "2         Fosl2   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "3          Relb   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "4          Plau   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "...         ...        ...                               ...\n",
       "7684        Gcg   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7685       Pdx1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7686       Ins1   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7687       Ins2   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "7688       Srp9   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "\n",
       "[7551 rows x 3 columns]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate targets\n",
    "mouse_msigdb = dc.translate_net(msigdb, source='geneset', target='genesymbol', source_tax_id=9606, target_tax_id=10090)\n",
    "mouse_msigdb"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0e9cfadb",
   "metadata": {},
   "source": [
    "Note that when performing homology convertion we might gain or lose some genes from one organism to another.\n",
    "\n",
    "Let us try the fruit fly (`7227`) now:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "9403ea9d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>genesymbol</th>\n",
       "      <th>collection</th>\n",
       "      <th>geneset</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Hand</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CG12648</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>twi</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HLH54F</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>dl</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_TNFA_SIGNALING_VIA_NFKB</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6594</th>\n",
       "      <td>Dmel\\CG6484</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6598</th>\n",
       "      <td>G6P</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6599</th>\n",
       "      <td>amon</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6600</th>\n",
       "      <td>Cbp53E</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6601</th>\n",
       "      <td>Srp9</td>\n",
       "      <td>hallmark</td>\n",
       "      <td>HALLMARK_PANCREAS_BETA_CELLS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5868 rows × 3 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       genesymbol collection                           geneset\n",
       "0            Hand   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "1         CG12648   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "2             twi   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "3          HLH54F   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "4              dl   hallmark  HALLMARK_TNFA_SIGNALING_VIA_NFKB\n",
       "...           ...        ...                               ...\n",
       "6594  Dmel\\CG6484   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6598          G6P   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6599         amon   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6600       Cbp53E   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "6601         Srp9   hallmark      HALLMARK_PANCREAS_BETA_CELLS\n",
       "\n",
       "[5868 rows x 3 columns]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Translate targets\n",
    "fly_msigdb = dc.translate_net(msigdb, source='geneset', target='genesymbol', source_tax_id=9606, target_tax_id=7227)\n",
    "fly_msigdb"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "test_decoupler",
   "language": "python",
   "name": "test_decoupler"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
